src/coreclr/jit/promotion.cpp

```diff
@@ -743,31 +783,551 @@ class ReplaceVisitor : public GenTreeVisitor<ReplaceVisitor>
     //   asg - The assignment
     //   user - The user of the assignment.
     //
-    void DecomposeAssignment(GenTreeOp* asg, GenTree* user)
+    void DecomposeAssignment(GenTree** use, GenTree* user)
     {
-        // TODO-CQ: field-by-field copies and inits.
+        GenTreeOp* asg = (*use)->AsOp();
 
-        if (asg->gtGetOp2()->OperIs(GT_LCL_VAR, GT_LCL_FLD))
+        if (!asg->gtGetOp1()->TypeIs(TYP_STRUCT))
         {
-            GenTreeLclVarCommon* rhsLcl = asg->gtGetOp2()->AsLclVarCommon();
-            if (rhsLcl->TypeIs(TYP_STRUCT))
+            return;
+        }
+
+        GenTree* dst = asg->gtGetOp1();
+        assert(!dst->OperIs(GT_COMMA));
+
+        GenTree* src = asg->gtGetOp2()->gtEffectiveVal();
+
+        Replacement* dstFirstRep             = nullptr;
+        Replacement* dstEndRep               = nullptr;
+        bool         dstInvolvesReplacements = asg->gtGetOp1()->OperIs(GT_LCL_VAR, GT_LCL_FLD) &&
+                                       OverlappingReplacements(dst->AsLclVarCommon(), &dstFirstRep, &dstEndRep);
+        Replacement* srcFirstRep             = nullptr;
+        Replacement* srcEndRep               = nullptr;
+        bool         srcInvolvesReplacements = asg->gtGetOp2()->OperIs(GT_LCL_VAR, GT_LCL_FLD) &&
+                                       OverlappingReplacements(src->AsLclVarCommon(), &srcFirstRep, &srcEndRep);
+
+        if (!dstInvolvesReplacements && !srcInvolvesReplacements)
+        {
+            return;
+        }
+
+        JITDUMP("Processing block operation [%06u] that involves replacements\n", Compiler::dspTreeID(asg));
+
+        if (dstInvolvesReplacements && (src->OperIs(GT_LCL_VAR, GT_LCL_FLD, GT_BLK, GT_FIELD) || src->IsConstInitVal()))
+        {
+            StatementList result;
+            EliminateCommasInBlockOp(asg, &result);
+
+            if (dstInvolvesReplacements && srcInvolvesReplacements)
+            {
+                JITDUMP("Copy [%06u] is between two physically promoted locals with replacements\n",
+                        Compiler::dspTreeID(asg));
+                JITDUMP("*** Conservative: Phys<->phys copies not yet supported; inserting conservative write-back\n");
```

Comment:
Is this case uncommon or something you want to work on later?

Would it be interesting to check for self assignment or do we just not see those?

Reply:
I want to handle sources in an upcoming change (I have most of this code written). Will keep in mind to add a check for self assignments then.